Include additional information for ShardSearchFailure exceptions

### Description

8.12

It will be helpful for `ShardSearchFailure` exceptions to include additional information such as the impacted index name, shard number and the offending query.

```
[2024-09-23T00:02:21,248][WARN ][org.elasticsearch.action.search.TransportSearchAction] [instance-0000000008] TransportSearchAction shard failure (partial results response)
org.elasticsearch.action.search.ShardSearchFailure: null
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:552) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:476) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.innerOnResponse(AbstractSearchAsyncAction.java:323) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.SearchActionListener.onResponse(SearchActionListener.java:33) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.SearchActionListener.onResponse(SearchActionListener.java:18) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.SearchExecutionStatsCollector.onResponse(SearchExecutionStatsCollector.java:62) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.SearchExecutionStatsCollector.onResponse(SearchExecutionStatsCollector.java:26) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:49) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleResponse(SearchTransportService.java:544) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleResponse(TransportService.java:1709) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1425) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:433) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:382) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundHandler.executeResponseHandler(InboundHandler.java:147) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:122) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:96) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:825) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:124) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:96) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:61) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.transport.netty4.Netty4MessageInboundHandler.channelRead(Netty4MessageInboundHandler.java:48) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[?:?]
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[?:?]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1383) ~[?:?]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1246) ~[?:?]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1295) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468) ~[?:?]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[?:?]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[?:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[?:?]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[?:?]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[?:?]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[?:?]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:689) ~[?:?]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:652) ~[?:?]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[?:?]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[?:?]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.time.DateTimeException: Field Year cannot be printed as the value -292275055 exceeds the maximum print width of 4
	at java.time.format.DateTimeFormatterBuilder$NumberPrinterParser.format(DateTimeFormatterBuilder.java:2926) ~[?:?]
	at java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(DateTimeFormatterBuilder.java:2529) ~[?:?]
	at java.time.format.DateTimeFormatterBuilder$CompositePrinterParser.format(DateTimeFormatterBuilder.java:2529) ~[?:?]
	at java.time.format.DateTimeFormatter.formatTo(DateTimeFormatter.java:1905) ~[?:?]
	at java.time.format.DateTimeFormatter.format(DateTimeFormatter.java:1879) ~[?:?]
	at org.elasticsearch.common.time.JavaDateFormatter.format(JavaDateFormatter.java:284) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.search.DocValueFormat$DateTime.format(DocValueFormat.java:280) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.search.DocValueFormat$DateTime.format(DocValueFormat.java:215) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.search.SearchSortValuesAndFormats.<init>(SearchSortValuesAndFormats.java:35) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.BottomSortValuesCollector.consumeTopDocs(BottomSortValuesCollector.java:66) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.onShardResult(SearchQueryThenFetchAsyncAction.java:115) ~[elasticsearch-8.12.0.jar:?]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.innerOnResponse(AbstractSearchAsyncAction.java:321) ~[elasticsearch-8.12.0.jar:?]
	... 46 more
```

--- a/docs/changelog/147566.yaml
+++ b/docs/changelog/147566.yaml
@@ -1,0 +1,6 @@
+area: Search
+issues:
+ - 113489
+pr: 147566
+summary: Include index and shard id in shard failure exception message
+type: enhancement

--- a/server/src/main/java/org/elasticsearch/action/ShardOperationFailedException.java
+++ b/server/src/main/java/org/elasticsearch/action/ShardOperationFailedException.java
@@ -75,4 +75,13 @@ public abstract class ShardOperationFailedException extends Exception implements
     public final Throwable getCause() {
         return cause;
     }
+
+    /**
+     * Returns a short message describing the failed shard.
+     * Subclasses that carry richer context (for example a node id) may override this further.
+     */
+    @Override
+    public String getMessage() {
+        return "shard [" + (index == null ? "_na" : index) + "][" + shardId + "]";
+    }
 }

--- a/server/src/test/java/org/elasticsearch/action/ShardOperationFailedExceptionTests.java
+++ b/server/src/test/java/org/elasticsearch/action/ShardOperationFailedExceptionTests.java
@@ -65,6 +65,29 @@ public class ShardOperationFailedExceptionTests extends ESTestCase {
         new Failure(null, randomInt(), randomAlphaOfLengthBetween(5, 10), randomFrom(RestStatus.values()), new IllegalArgumentException());
     }
 
+    public void testGetMessageIncludesIndexAndShard() {
+        Failure failure = new Failure(
+            "books",
+            7,
+            randomAlphaOfLengthBetween(5, 10),
+            randomFrom(RestStatus.values()),
+            new IllegalArgumentException()
+        );
+        assertEquals("shard [books][7]", failure.getMessage());
+        assertEquals("org.elasticsearch.action.ShardOperationFailedExceptionTests$Failure: shard [books][7]", failure.toString());
+    }
+
+    public void testGetMessageWhenIndexIsNull() {
+        Failure failure = new Failure(
+            null,
+            -1,
+            randomAlphaOfLengthBetween(5, 10),
+            randomFrom(RestStatus.values()),
+            new IllegalArgumentException()
+        );
+        assertEquals("shard [_na][-1]", failure.getMessage());
+    }
+
     private static class Failure extends ShardOperationFailedException {
 
         Failure(@Nullable String index, int shardId, String reason, RestStatus status, Throwable cause) {

